### PR TITLE
Cyptography pr 20566 rebase

### PR DESCRIPTION
**Pull request by abadger**
##### SUMMARY
Port the ansible code to work with the cryptography python library as this is the library that everyone in the python community is getting behind for cryptographic functions.

We maintain a fallback for pycrypto so that older distros will still work out of the box.

The idea is for the following to work:
* ```pip install ansible``` to install with cryptography as its setuptools dep (install time and runtime)
* ```ANSIBLE_CRYPTO_BACKEND=pycrypto pip install ansible``` will install with pycrypto as its setuptools  dep (installtime and runtime).
* If both cryptography and pycrypto are installed, ansible will prefer to use cryptography whether or not it is the setuptools dep.

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
primarily vault

##### ANSIBLE VERSION

```
devel (2.4)
```

##### ADDITIONAL INFORMATION
based on https://github.com/ansible/ansible/pull/24894

Which in turn is based on https://github.com/ansible/ansible/pull/20566

@reaperhulk, @alikins, Please have a look for the crypto code.  I'll see if we can wrangle some security people to look at it as well.

@mattclay, @gundalow Please have a look for the test portion.  docker images probably need to be updated.  It would be good to find a way to test the code when using both cryptography and when using pycrypto.  We might be able to do that by monkeypatching HAS_CRYPTOGRAPHY to False when we want to choose the pycrypto code (This is assuming that the test machine has both installed).

### CODE REVIEW

**Comment by abadger** [comment]
Okay, integration tests passing locally (on Fedora 25) with:
```
 test/runner/ansible-test integration  --python 3.5 vault
 test/runner/ansible-test integration  --python 2.7 vault
```

unittests passing locally with
```
test/runner/ansible-test units --python 3.5 
test/runner/ansible-test units --python 2.7
```

The unittests require that both pycrypto and cryptography are installed which is not the case in the docker images.  what do we want to do about that? Options:

* Install both cryptography and pycrypto
* Rewrite the unittests to require only cryptography and update the docker images (easy to rewrite as it's just removing some calls that specifically target pycrypto)
* Have certain unittests skip if pycrypto isn't installed.

**Comment by abadger** [comment]
Okay, unittests now are duplicated for cryptography and pycrypto as their backend and only run condtionally depending on which is installed.  @reaperhulk, I also found out why you had to add paramiko when we didn't.  cryptography is using pkg_resources at runtime.  Because of that, it forces our deps to be available too.  Since paramiko wasn't available, we go boom.  pkg_resources is unfortunately fragile like this (we have to use it too but for a different reason).  Since you're only using it for loading plugins, you might want to look into something that's more robust like straight.plugins.

File: lib/ansible/modules/cloud/amazon/ec2_win_password.py
Comment (reaperhulk) [inline lib/ansible/modules/cloud/amazon/ec2_win_password.py]: This will fail on Python 3 as f is opened as 'r' rather than 'rb' and load_pem_private_key requires bytes. key_passphrase must also be bytes -- if that path accepts utf8 `encode('utf8')` would be appropriate

File: test/integration/targets/vault/vault-password-wrong
Comment (reaperhulk) [inline test/integration/targets/vault/vault-password-wrong:1]: 😆

File: lib/ansible/parsing/vault/__init__.py
Comment (reaperhulk) [inline lib/ansible/parsing/vault/__init__.py:783]: CTR mode doesn't actually require padding, so this is unnecessary. That said, I assume you're staying compatible with existing vault implementations which already do this. It's not a security thing, just a few wasted bytes/CPU cycles.

**Review by reaperhulk — COMMENTED** [review]
Thanks for taking this over and doing so much work! I've put a few comments on it after review.

One thing that might be useful to know: cryptography is planning on beginning to ship a manylinux1 wheel soon(-ish). When that happens ansible users on manylinux1 compatible distributions will no longer need to install OpenSSL development headers. cffi already ships a manylinux1 wheel, which, if the user is on pip 8.1+, obviates the need for libffi dev headers.

File: lib/ansible/parsing/vault/__init__.py
Comment (jimi-c) [inline lib/ansible/parsing/vault/__init__.py]: Is this really required?

File: lib/ansible/parsing/vault/__init__.py
Comment (bcoca) [reply lib/ansible/parsing/vault/__init__.py:783]: yes, we've been wanting to change that for a while but were waiting until we made the architecture more pluggable to allow for old/new formats to be used transparently.

File: lib/ansible/parsing/vault/__init__.py
Comment (abadger) [reply lib/ansible/parsing/vault/__init__.py]: It shouldn't be.... But I felt like I needed to put something in here because we (alikins and I) think that all of these should be caught in __init__().  Not sure how we can signal this otherwise....

File: lib/ansible/modules/cloud/amazon/ec2_win_password.py
Comment (abadger) [reply lib/ansible/modules/cloud/amazon/ec2_win_password.py]: Good point.  I'll convert this to use 'rb'.

**Comment by abadger** [comment]
Thanks @jimi-c @bcoca for the reviews.  @reaperhulk, thanks for the review and the bulk of the code!  The information about cryptography gaining a manylinux wheel soon should also be good news for us going forward.

I think that I've addressed all the issues you've collectively brought up. The one outstanding question I see is whether the integration tests will work with just the changes here once the docker images are updated.  I'm hopeful since integration tests are passing here but we shall have to see.  I'm also hoping that mckerrj and I can find someone on the redhat security team to review this code before we put it in but even if that doesn't happen, a post-merge review should still be doable.

I'm going on vacation followed by ansiblefest in London and will be gone for the next two weeks.  (Can check in on this sporadically, though). I have no objections if anyone else wants to work on this for any furthe rissues that come up in that time.

File: hacking/README.md
Comment (nitzmahone) [inline hacking/README.md:19]: Unrelated, but wonder if we should ditch reference to easy_install, as pip now recommends `get-pip.py`...

File: lib/ansible/parsing/vault/__init__.py
Comment (nitzmahone) [inline lib/ansible/parsing/vault/__init__.py]: Minor consistency nit- our stuff seems to be using `[action]_[backend]()`, where these aliased imports have that reversed. I'd vote for `[action]_[backend]()` over this- makes the code a little easier to read...

File: lib/ansible/parsing/vault/__init__.py
Comment (nitzmahone) [inline lib/ansible/parsing/vault/__init__.py:94]: Do we know a minimum version we could specify?

**Comment by gundalow** [comment]
Docker images have been updated, closing and re-opening to rerun CI

**Review by nitzmahone — COMMENTED** [review]
LGTM, couple of minor nits that may or may not be worth addressing...

File: hacking/README.md
Comment (gundalow) [reply hacking/README.md:19]: This may need changing in other places, such as install guide.

Comment (abadger) [reply hacking/README.md:19]: I think we should do this one separately.  There's questions of  availability that we'd need to figure out. (I checked whether I had get_pip.py on a centos6 box and it wasn't there... then I checked fedora25 and it wasn't installed either).

File: lib/ansible/parsing/vault/__init__.py
Comment (abadger) [reply lib/ansible/parsing/vault/__init__.py:94]: It may be pycrypto 2.6 but I'm not sure.  2.6 is the minimum version that was specified in the requirements.txt file (in code in setup.py after this patch).  If you want to do some research and update the patch.... The message as written isn't a regression, though...

**Review by nitzmahone — APPROVED** [review]

**Comment by abadger** [comment]
Failures and what's probably wrong:
* freebsd-10; looks like the instance went away, not the PRs fault.
* ubuntu 14, 16, 16/py3 all need to have gcc added to the docker images.

**Comment by mattclay** [comment]
The docker images have been updated and CI is now passing.

File: hacking/README.md
Comment (abadger) [reply hacking/README.md:19]: Okay, on python-2.7.13 and python-3.5 (have to check older versions for pyhton2 but I'm pretty sure it wasn't in python-2.6.x) you can run this:
```
$ python -m ensurepip
```

to get pip and dependencies installed.

Comment (reaperhulk) [reply hacking/README.md:19]: ensurepip was added in 3.4+ and backported to 2.7 in 2.7.9+

**Comment by abadger** [comment]
/me will merge tomorrow when I'll be around all day in case some unexpected error arises from this.

**Comment by abadger** [comment]
Merged to devel for 2.4.0 inclusion

**Comment by abadger** [comment]
Thanks to @reaperhulk @alikins @nbraud and the many reviewers and testers for bringing this PR to completion!

**Pull request merged**
